minetest/minetest
# Decouple liquid pointing from liquidtype

This PR introduces a new node property called `liquid_raycast`. It takes one of 3 values:

* `true`: Node is considered to be a liquid for raycasting/pointing purposes. This means, RayCasts will consider this node to be a liquid, and items with `liquids_pointable=true`, too
* `false`: This is not the case
* `nil` (default). Automatic. If `liquidtype ~= "none"`, this is the same as `true`, otherwise `false`

This is a change from the current behavior. Currently, nodes are considered to be liquids for raycasting/pointing if `liquidtype ~= "none"`.

Note that the default value `nil` leads to exactly the current behavior, so existing nodes will not change in their behavior. This PR should be 100% backwards-compatible.

This PR will have the following benefits:

* You can now make liquid nodes that are not pointable at all (not even MTG buckets)
* You can now make nodes pointable by MTG buckets even if they don't have liquid flowing physics.

## Rationale

Basically, this PR is a part of my quest to decouple the various features of liquids into distinct, independent features. It is a contination of #10810.

I dislike the current behavior because it strongly couples the pointing behavior with liquid flow physics. E.g. setting `liquidtype` does two different things at once: 1) Enable liquid flowing physics and 2) change how raycasting/pointing works. Thus, `liquidtype` strongly couples two different things into one. This means that the two benefits I listed above are currently impossible, thus limiting flexibility.

## To do

This PR is Ready for Review.

## How to test

Test 1:
In DevTest, give yourselves `testnodes:fake_liquid_raycast` and place it. Point at this node with the hand. It should not be pointed. Point with it with the Liquids Pointer tool (`testtools:liquids_pointer`) or bucket. Now it should be pointed.

Test 2:
Give yourselves `testnodes:liquid_noraycast` and place it. It should be impossible to point it with any tool (both the source and flowing node), including the Liquids Pointer or Bucket.

Test 3:
Test if pointing still works as expected with the other nodes (both liquids and non-liquids).

Test 4:
Test with RayCast functions from Lua. (I did not do this myselves yet.)
